lib/Containers/MerkleTree.cpp

```diff
@@ -759,41 +759,42 @@ MerkleTree<Hasher, BranchingBits>::diff(
     tree2 = dummy;
   }
   // Now the rangeMin of tree1 is <= the rangeMin of tree2.
-
   TRI_ASSERT(tree1->meta().rangeMin <= tree2->meta().rangeMin);
 
+  TRI_ASSERT(tree1->numberOfShards() == tree2->numberOfShards());
+
   std::vector<std::pair<std::uint64_t, std::uint64_t>> result;
   std::uint64_t n = nodeCountAtDepth(depth);
 
   auto addRange = [&result](std::uint64_t min, std::uint64_t max) {
-    if (result.size() == 0) {
-      result.push_back(std::make_pair(min, max));
-      return;
-    }
-    if (result.back().second + 1 == min) {
+    if (!result.empty() && result.back().second + 1 == min) {
       // Extend range: //
       result.back().second = max;
-      return;
+    } else {
+      result.push_back(std::make_pair(min, max));
     }
-    result.push_back(std::make_pair(min, max));
   };
 
   // First do the stuff tree2 does not even have:
   std::uint64_t keysPerBucket =
       (tree1->meta().rangeMax - tree1->meta().rangeMin) / n;
   std::uint64_t index1 = 0;
   std::uint64_t pos = tree1->meta().rangeMin;
-  for (; pos < tree2->meta().rangeMin; pos += keysPerBucket) {
+  for (; pos < tree2->meta().rangeMin && pos < tree1->meta().rangeMax;
+       pos += keysPerBucket) {
     Node const& node1 = tree1->node(index1);
     if (node1.count != 0) {
       addRange(pos, pos + keysPerBucket - 1);
     }
     ++index1;
   }
   // Now the buckets they both have:
+  TRI_ASSERT(pos <= tree2->meta().rangeMin);
+  // note that pos can be < tree2->meta().rangeMin if the trees do not overlap
+  // at all
```

Comment:
We can strengthen this assertion:
```suggestion
  TRI_ASSERT(pos == tree2->meta().rangeMin || (pos == tree1->meta().rangeMax && tree2->meta().rangeMin > tree1->meta().rangeMax));
  // if the trees do not overlap, then the previous loop stops at tree1.rangeMax already
```